fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,383 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Directory implementation backed by FDB which attempts to
+ * model a file system on top of FoundationDB.
+ *
+ * @see <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/store/Directory.html">Directory</a>
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FDBDirectory extends Directory {
+    private static final Logger LOG = LoggerFactory.getLogger(FDBDirectory.class);
+    public static final int DEFAULT_BLOCK_SIZE = 16_384;
+    private final AtomicLong nextTempFileCounter = new AtomicLong();
+    public Transaction txn;
+    public final Subspace subspace;
+    private final Subspace metaSubspace;
+    private final Subspace dataSubspace;
+    private final byte[] sequenceSubspaceKey;
+
+    private final LockFactory lockFactory;
+    private final int blockSize;
+    private static final Set<String> EMPTY_SET = new HashSet<>();
+    private AtomicLong atomicLong;
+    private boolean hasWritten;
+    private CompletableFuture<Void> setIncrement;
+    private final Cache<String, FDBLuceneFileReference> fileReferenceCache;
+    private final Cache<Pair<Long, Integer>, CompletableFuture<byte[]>> blockCache;
+    private final Map<String, Long> reads;
+
+    public FDBDirectory(Subspace subspace, Transaction txn) {
+        this(subspace, txn, NoLockFactory.INSTANCE);
+    }
+
+    FDBDirectory(Subspace subspace, Transaction txn, LockFactory lockFactory) {
+        this(subspace, txn, lockFactory, DEFAULT_BLOCK_SIZE);
+    }
+
+    FDBDirectory(Subspace subspace, Transaction txn, LockFactory lockFactory, int blockSize) {
+        assert subspace != null;
+        assert txn != null;
+        assert lockFactory != null;
+        this.txn = txn;
+        this.subspace = subspace;
+        final Subspace sequenceSubspace = subspace.subspace(Tuple.from("s"));
+        this.sequenceSubspaceKey = sequenceSubspace.pack();
+        this.metaSubspace = subspace.subspace(Tuple.from("m"));
+        this.dataSubspace = subspace.subspace(Tuple.from("d"));
+        this.lockFactory = lockFactory;
+        this.blockSize = blockSize;
+        this.fileReferenceCache = CacheBuilder.newBuilder().initialCapacity(128).maximumSize(1024).recordStats().build();
+        this.blockCache = CacheBuilder.newBuilder().concurrencyLevel(16).initialCapacity(128).maximumSize(1024).recordStats().build();
+        this.reads = new ConcurrentHashMap<>();
+    }
+
+    /**.
+     * Sets increment if not set yet.
+     * @return Future representing the status of setting the increment
+     */
+    private CompletableFuture<Void> setIncrement() {
+        return txn.get(sequenceSubspaceKey).thenAcceptAsync(
```

Comment:
don't know that it matters, but I don't think this is thread-safe, since `atomicLong` isn't a volatile or atomic variable. If two threads attempt to call `setIncrement()` simultaneously, you could end up setting `atomicLong` to two separate objects; that might make the sequence count be off by one under concurrent load.

Reply:
This is at startup but I will wrap in synchronized.  Thanks.

Reply:
Fixing

Reply:
I just cleaned this up and simplified.